clash-prelude/src/Clash/Class/BitPack/BitIndex.hs

```diff
@@ -0,0 +1,148 @@
+{-|
+Copyright  :  (C) 2013-2016, University of Twente
+License    :  BSD2 (see the file LICENSE)
+Maintainer :  Christiaan Baaij <christiaan.baaij@gmail.com>
+-}
+
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE TypeFamilies #-}
+
+{-# LANGUAGE Trustworthy #-}
+
+{-# OPTIONS_HADDOCK show-extensions #-}
+
+module Clash.Class.BitPack.BitIndex where
+
+import GHC.TypeLits                   (KnownNat, type (+), type (-))
+
+import Clash.Class.BitPack.Internal   (BitPack (..))
+import Clash.Promoted.Nat             (SNat (..))
+import Clash.Sized.Internal.BitVector (BitVector, Bit, index#, lsb#, msb#,
+                                       replaceBit#, setSlice#, slice#, split#)
+
+{- $setup
+>>> :set -XDataKinds
+>>> import Clash.Prelude
+-}
+
+{-# INLINE (!) #-}
+-- | Get the bit at the specified bit index.
+--
+-- __NB:__ Bit indices are __DESCENDING__.
+--
+-- >>> pack (7 :: Unsigned 6)
+-- 00_0111
+-- >>> (7 :: Unsigned 6) ! 1
+-- 1
+-- >>> (7 :: Unsigned 6) ! 5
+-- 0
+-- >>> (7 :: Unsigned 6) ! 6
+-- *** Exception: (!): 6 is out of range [5..0]
+-- ...
+(!) :: (BitPack a, Enum i) => a -> i -> Bit
+(!) v i = index# (pack v) (fromEnum i)
+
+{-# INLINE slice #-}
+-- | Get a slice between bit index @m@ and and bit index @n@.
+--
+-- __NB:__ Bit indices are __DESCENDING__.
+--
+-- >>> pack (7 :: Unsigned 6)
+-- 00_0111
+-- >>> slice d4 d2 (7 :: Unsigned 6)
+-- 001
+-- >>> slice d6 d4 (7 :: Unsigned 6)
+-- <BLANKLINE>
+-- <interactive>:...
+--     • Couldn't match type ‘7 + i0’ with ‘6’
+--         arising from a use of ‘slice’
+--       The type variable ‘i0’ is ambiguous
+--     • In the expression: slice d6 d4 (7 :: Unsigned 6)
+--       In an equation for ‘it’: it = slice d6 d4 (7 :: Unsigned 6)
+slice :: (BitPack a, BitSize a ~ ((m + 1) + i)) => SNat m -> SNat n -> a
+      -> BitVector (m + 1 - n)
```

Comment:
I know this is probably lifted as-is from somewhere else, but I would take the time to fix the formatting of that type signature while the diff is all green anyway